### ISSUE TITLE
Client with http channel broken, after 'connection refused' error

### ISSUE THREAD

**Issue by pbedat** [creachadair/jrpc2#102]
After performing a single request against a http server, that is not running, the client seems to be in an invalid state, where all further requests fail with the same "connection refused" error.

I've modified your code example from the docs to demonstrate the behaviour:

```go
package shop

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"runtime"
	"strings"
	"testing"

	"github.com/creachadair/jrpc2"
	"github.com/creachadair/jrpc2/handler"
	"github.com/creachadair/jrpc2/jhttp"
)

func TestWat(t *testing.T) {

	b := jhttp.NewBridge(handler.Map{
		"Test": handler.New(func(ctx context.Context, ss []string) string {
			return strings.Join(ss, " ")
		}),
	}, nil)
	defer b.Close()

	hsrv := http.Server{
		Handler: b,
		Addr:    ":1234",
	}
	go hsrv.ListenAndServe()

	runtime.Gosched()

	// Set up a client using an HTTP channel, and use it to call the test
	// service exported by the bridge.
	ch := jhttp.NewChannel("http://127.0.0.1:1234", nil)
	cli := jrpc2.NewClient(ch, nil)

	var result string
	if err := cli.CallResult(context.Background(), "Test", []string{
		"full", "plate", "and", "packing", "steel",
	}, &result); err != nil {
		t.Fatalf("Call failed: %v", err)
	}

	hsrv.Close()

	if err := cli.CallResult(context.Background(), "Test", []string{
		"full", "plate", "and", "packing", "steel",
	}, &result); err == nil {
		t.Fatal("expected err")
	}

	hsrv = http.Server{
		Handler: b,
		Addr:    ":1234",
	}
	go hsrv.ListenAndServe()

	runtime.Gosched()

	if err := cli.CallResult(context.Background(), "Test", []string{
		"full", "plate", "and", "packing", "steel",
	}, &result); err != nil {
		log.Fatalf("Call after failed call failed: %v", err)
	}

	fmt.Println("Result:", result)
}
```

After some debugging, I found this line:

https://github.com/creachadair/jrpc2/blob/25bf5bc73e326f786ed9400d3ee2c34777d55023/client.go#L227

So I'm not quite sure if this behaviour is really intended?

Shouldn't the error be cleared after it has been reported to the caller?

**Comment by creachadair** [comment]
Thanks for the report. I agree, the client probably shouldn't treat HTTP errors as persistent.

**Comment by creachadair** [comment]
I did a little investigation, and I think the easiest way to deal with this situation is to plug in an HTTP client that can handle retries, e.g., [go-retryablehttp](https://github.com/hashicorp/go-retryablehttp):

```go
rc := retryablehttp.NewClient()
rc.RetryWaitMin = 5*time.Millisecond
rc.RetryWaitMax = 100*time.Millisecond
rc.RetryMax = 10

ch := jhttp.NewChannel("http://example.com/rpc", &jhttp.ChannelOptions{
   Client: rc,
})

cli := jrpc2.NewClient(ch, opts)
```

**Comment by pbedat** [comment]
> I don't want to bake HTTP-specific knowledge into the client

I dont think it's http specific. Let's assume a rpc request causes a panic in the server and another request is issued while the service is recovering. This will also lead to the persistent error state.

> but we do want a way to express the idea of "an error that is transient and ought to be retried".
Maybe some callback option, that identifies errors as transient. But then you'd have to carry the response id around to finish the pending request. And this would make the current channel interfaces much more complicated :/

> My intuition is that the channel probably ought to deal with retries in some way.
Yeah, as I'm trying to work around the problem, I also thought about wrapping the channel in some way. Some kind of generic retry wrapper would be nice.

Edit: Just saw your suggestion to use retryablehttp. This is nice! Thanks for this

Another way to solve it would be a `func OnStop(err error)` callback. This would allow you to recreate the client, when OnStop is called with an error (other than errClientStopped).

**Comment by creachadair** [comment]
> I dont think it's http specific. Let's assume a rpc request causes a panic in the server and another request is issued while the service is recovering. This will also lead to the persistent error state.

I don't think the channel is the right place to recover from a panic in a handler. The choice about whether it is possible to recover without corrupting the state of the server is up to the handler to decide.

Even leaving aside panics, once a channel fails the right way to recover is to establish a new channel. If a channel can heal itself that's fine, but I don't think the client should be involved in the choice: either the channel recovers on its own or the user of the client needs to make a choice about the right way to recover.

> Maybe some callback option, that identifies errors as transient. But then you'd have to carry the response id around to finish the pending request. And this would make the current channel interfaces much more complicated :/ 

Yes, I don't think that complexity makes sense here.

> Yeah, as I'm trying to work around the problem, I also thought about wrapping the channel in some way. Some kind of generic retry wrapper would be nice.

You could pretty easily write your own Channel wrapper that handles certain errors and does recovery.

> Another way to solve it would be a `func OnStop(err error)` callback. This would allow you to recreate the client, when OnStop is called with an error (other than errClientStopped).

Sure. And you don't need a fork for that, it can live entirely inside the channel:

```go
type StopChannel struct{
    C channel.Channel
    OnError func(*StopChannel, error) error
}

func (s *StopChannel) Recv() ([]byte, error) {
   msg, err := s.C.Recv()
   if err != nil {
      err = s.OnError(s, err)
      if err != nil {
         return nil, err
      }
      return s.Recv()
   }
   return msg, nil
}

// etc.
```

Or words to that effect.

**Comment by creachadair** [comment]
Tangentially, it's worth noting that HTTP is kind of a weird corner case here. A protocol like JSON-RPC is intended for situations where you don't have (or don't want) the full semantics of HTTP, typically pipes between processes or local sockets. If you already have support for HTTP between the client and the server, there is not much reason to use JSON-RPC at all-HTTP already supports framing, request isolation, metadata, side channels, and so on. 

The jhttp package is kind of a hack to bridge between them, but in a situation where the HTTP part is load bearing, I'd recommend you consider moving away from JSON-RPC entirely. When everyone is already using HTTP anyway, it's easier for both clients and servers to ditch the extra framing that JSON-RPC uses to deal with channels lacking that structure.

**Comment by pbedat** [comment]
The StopChannel wrapper works like a charm! Thanks for pointing me at this.

But let's leave the http bridge aside for a moment (I just used it because it was convenient and performance is not a problem).

When we cross process or network boundaries, regardless of the underlying transport (TCP, WebSockets, HTTP, etc.), segmentation will happen from time to time. In this case, the client would be stopped, when you don't handle the segmentation e.g. with the StopChannel. This situation can be really problematic, when a client becomes inresponsive in production, an the errors of the client are then misleading (because it repeats the initial error).

Maybe I'm overlooking something, but don't you think it would be better to document this behaviour, or offer wrappers like the above?

**Comment by pbedat** [comment]
Really, this is not about http :)

I think the documentation could be improved by either making this failure state part of the public interface of the client or by documenting it in the Call* methods.

When I'm integrating a module, after figuring out how to use it, I'm looking for failure states, that I have to handle. At no point I was aware, that the jrpc2.Client would enter a "stopped" state, that I have to handle.

Now if the client would wrap the errors, that lead to a stopped state with a more specific error or the client would receive a "Stopped" field, it would become evident, that I as a user of the client, must somehow handle this situation.
Or you introduce the StopChannel helper, and list an example, where a tcp channel is redialed after disconnect.

Hope you understand what I'm trying to say :)

**Comment by pbedat** [comment]
Allright! Thank you for your help, patience and the insightful discussion. While my "issue" has been resolved, I leave it to you, if you want to close this issue.
I'd also be glad to help with feedback if you are in need of it.

**Comment by creachadair** [comment]
> Allright! Thank you for your help, patience and the insightful discussion. While my "issue" has been resolved, I leave it to you, if you want to close this issue.
> 
> I'd also be glad to help with feedback if you are in need of it.

Thanks, I appreciate the offer. I've been experimenting with a couple different ways to make this case easier to deal with, and I'll probably tag you on a PR when I make up my mind.

**Comment by pbedat** [comment]
Absolutely! Left a comment at the PR

**Comment by creachadair** [comment]
Thanks. #103 is now merged and the change is available at and after the v1.1.0 tag.
